### PR TITLE
Freshness pass and bug fixes: Nullable reference types

### LINKED ISSUE #21941: Arrays that contain reference types
I Google searched for "arrays of nullable reference types c#" because there are some interesting design decisions in this area of .NET. I found [this very informative article](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references) on the pitfalls of using nullable reference types. One of these pitfalls is arrays of non-nullable references. While the article is informative in general, there is only a paragraph of information about this particular topic and I feel like there could be more info because I'm currently working with these in my project and trying to work out what options there are.

My request isn't in the exact format you asked for, because I'm not asking for a whole article to be written on this. It would be nice to see a paragraph or two added to the existing article, discussing:

- Why arrays of non-nullable references are initialised to null.
- Why the compiler lets the programmer create an array of non-nullable references without a warning.
- What options you can take / recommended practices to avoid potential issues resulting from non-nullable references being set to null unexpectedly.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: a690704d-da32-dca7-ea3e-66b84fb1d00d
* Version Independent ID: 11f314de-dd3a-362e-b2bb-d0a2133cfc7b
* Content: [Nullable reference types](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references)
* Content Source: [docs/csharp/nullable-references.md](https://github.com/dotnet/docs/blob/main/docs/csharp/nullable-references.md)
* Product: **dotnet-csharp**
* Technology: **csharp-null-safety**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

---
[Associated WorkItem - 350756](https://dev.azure.com/msft-skilling/Content/_workitems/edit/350756)

### LINKED ISSUE #30348: Imagine My Surprise When Nullable Reference Types Became An Issue
I pay a hefty fee to Microsoft to subscribe to Visual Studio and have done so for a long time.  I installed the latest creation the other day and compiled some of my old code in a new project.  Imagine my surprise when the compiler complained about nullable reference types.   Microsoft declined to send me a memo regarding substantial deprecations of syntax.  In fact, I *NEVER* hear from Microsoft.  The trade press sometimes mentions Microsoft when 97 security threats are discovered, many of them zero days.  I am a software developer and have been for many decades.  My old company took great care in documenting product changes.  Of course, my old company had only thousands of customers, not more than a billion.  My next employer was based in Paris.  They hired "writers" who were not technical but did speak English.  They conversed with developers and put words on paper.  Extracting product design from their documents, if you could get through them, was impossible.  Apparently, my industry, which is your industry, hates to write documentation.  Writing code is so much more enjoyable.  This page is good.  I'm glad I found it.  The error messages that Visual Studio issues when it discovers a reference type, that may or may not be null at runtime, could, but doesn't, refer to it.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: a690704d-da32-dca7-ea3e-66b84fb1d00d
* Version Independent ID: 11f314de-dd3a-362e-b2bb-d0a2133cfc7b
* Content: [Nullable reference types](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references)
* Content Source: [docs/csharp/nullable-references.md](https://github.com/dotnet/docs/blob/main/docs/csharp/nullable-references.md)
* Product: **dotnet-csharp**
* Technology: **csharp-null-safety**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

---
[Associated WorkItem - 350754](https://dev.azure.com/msft-skilling/Content/_workitems/edit/350754)

### LINKED ISSUE #33667: question about ' Sets the nullable annotation context and nullable warning context to enable.
in your document you write :
#nullable enable: **Sets the nullable annotation context and nullable warning context to enable.**
I don't understand  because "nullable annotation context" and "nullable warning context " are mutually exclusive
ant their properties are conflicting : for example 
- _Dereference warnings_ is enabled with _nullable warning context_
- _Dereference warnings_ is disabled with _nullable annotation context_
---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: a690704d-da32-dca7-ea3e-66b84fb1d00d
* Version Independent ID: 11f314de-dd3a-362e-b2bb-d0a2133cfc7b
* Content: [Nullable reference types](https://learn.microsoft.com/en-us/dotnet/csharp/nullable-references#nullable-contexts)
* Content Source: [docs/csharp/nullable-references.md](https://github.com/dotnet/docs/blob/main/docs/csharp/nullable-references.md)
* Product: **dotnet-csharp**
* Technology: **csharp-null-safety**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

---
[Associated WorkItem - 350753](https://dev.azure.com/msft-skilling/Content/_workitems/edit/350753)

### LINKED ISSUE #36767: I was not able to find good info on nullable (reference type) arrays
My search: https://www.bing.com/search?q=c%23+nullable+array+elements

This is the best I could find in our docs, after a lot of clicking: https://learn.microsoft.com/en-us/dotnet/csharp/nullable-references#arrays

This is what I was looking for (since I always forget): https://stackoverflow.com/a/54010485

> There's a difference between the array object itself and the elements of the array. When you use `decimal?[]` you are declaring an array whose elements are nullable decimal values. If you were to use `decimal[]?` (I'm not sure that's valid syntax, but just for explanation purposes, assume it is) you would be declaring a variable that may refer to an array or may not refer to anything.

/cc @KathleenDollard @BillWagner 

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: a690704d-da32-dca7-ea3e-66b84fb1d00d
* Version Independent ID: 11f314de-dd3a-362e-b2bb-d0a2133cfc7b
* Content: [Nullable reference types](https://learn.microsoft.com/en-us/dotnet/csharp/nullable-references#arrays)
* Content Source: [docs/csharp/nullable-references.md](https://github.com/dotnet/docs/blob/main/docs/csharp/nullable-references.md)
* Product: **dotnet-csharp**
* Technology: **csharp-null-safety**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

---
[Associated WorkItem - 350752](https://dev.azure.com/msft-skilling/Content/_workitems/edit/350752)

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #21941: Add description of the default values for arrays of reference types.
Fixes #30348: Edit pass
Fixes #33667: Update the description of the nullable context to match the language used in the C# 8 draft standard
Fixes #36767: Add language on declaring a non-null array of nullable reference types and a nullable array of non-nullable reference types.

Also, perform a general edit pass.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/language-reference/builtin-types/arrays.md](https://github.com/dotnet/docs/blob/765be6b7538ab6415435b4e86e465e6ccdf9759a/docs/csharp/language-reference/builtin-types/arrays.md) | [Arrays](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/arrays?branch=pr-en-us-43914) |
| [docs/csharp/language-reference/compiler-options/language.md](https://github.com/dotnet/docs/blob/765be6b7538ab6415435b4e86e465e6ccdf9759a/docs/csharp/language-reference/compiler-options/language.md) | ["Compiler Options - language feature rules"](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/language?branch=pr-en-us-43914) |
| [docs/csharp/language-reference/keywords/where-generic-type-constraint.md](https://github.com/dotnet/docs/blob/765be6b7538ab6415435b4e86e465e6ccdf9759a/docs/csharp/language-reference/keywords/where-generic-type-constraint.md) | [where (generic type constraint) (C# Reference)](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/where-generic-type-constraint?branch=pr-en-us-43914) |
| [docs/csharp/language-reference/operators/null-forgiving.md](https://github.com/dotnet/docs/blob/765be6b7538ab6415435b4e86e465e6ccdf9759a/docs/csharp/language-reference/operators/null-forgiving.md) | [docs/csharp/language-reference/operators/null-forgiving](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving?branch=pr-en-us-43914) |
| [docs/csharp/nullable-references.md](https://github.com/dotnet/docs/blob/765be6b7538ab6415435b4e86e465e6ccdf9759a/docs/csharp/nullable-references.md) | [Nullable reference types](https://review.learn.microsoft.com/en-us/dotnet/csharp/nullable-references?branch=pr-en-us-43914) |

### CODE REVIEW

**Comment by BillWagner** [comment]
ping @adegeo for review

File: docs/csharp/language-reference/builtin-types/arrays.md [hunk 1]
```diff
@@ -26,6 +26,24 @@ You can store multiple variables of the same type in an array data structure. Yo
 type[] arrayName;
 ```
 
+An array is a reference type, so the array can be a nullable reference type. The element types might be reference types, so an array can be declared to hold nullable reference types. The following example declarations show the different syntax used to declare the nullability of the array or the elements:
```
Comment (adegeo) [inline docs/csharp/language-reference/builtin-types/arrays.md:29]: Good place for a link 
```suggestion
An array is a reference type, so the array can be a [nullable reference](../nullable-references.md) type. The element types might be reference types, so an array can be declared to hold nullable reference types. The following example declarations show the different syntax used to declare the nullability of the array or the elements:
```

File: docs/csharp/nullable-references.md [hunk 2]
```diff
@@ -209,32 +201,36 @@ You can specify different behavior using [constraints](programming-guide/generic
 
 These constraints help provide more information to the compiler on how `T` is used. That helps when developers choose the type for `T` and provides better *null-state* analysis when an instance of the generic type is used.
 
-## Nullable contexts
+## Nullable context
+
+The *nullable context* determines how nullable reference type annotations are handled and what warnings are produced by static null state analysis. The nullable context contains two flags: the *annotation* setting and the *warning* setting.
+
+Both the *annotation* and *warning* settings are disabled by default for existing projects. Starting in .NET 6, both flags are enabled by default for *new* projects. The reason for two distinct flags for the nullable context is to make it easier to migrate large projects that predate the introduction of nullable reference types.
```
Comment (adegeo) [inline docs/csharp/nullable-references.md:208]: Should this instead be the C# version, not the .NET version?

**Review by adegeo — APPROVED** [review]
LGTM! Few suggestions and approved to let you decide.

File: docs/csharp/nullable-references.md (hunk 2 above)
Comment (BillWagner) [inline docs/csharp/nullable-references.md:208]:
```suggestion
Both the *annotation* and *warning* settings are disabled by default for existing projects. Starting in .NET 6 (C# 10), both flags are enabled by default for *new* projects. The reason for two distinct flags for the nullable context is to make it easier to migrate large projects that predate the introduction of nullable reference types.
```

File: docs/csharp/language-reference/builtin-types/arrays.md [hunk 3]
```diff
@@ -26,6 +26,24 @@ You can store multiple variables of the same type in an array data structure. Yo
 type[] arrayName;
 ```
 
+An array is a reference type, so the array can be a [nullable reference](../nullable-references.md) type. The element types might be reference types, so an array can be declared to hold nullable reference types. The following example declarations show the different syntax used to declare the nullability of the array or the elements:
```
Comment (BillWagner) [inline docs/csharp/language-reference/builtin-types/arrays.md:29]:
```suggestion
An array is a reference type, so the array can be a [nullable reference](../../nullable-references.md) type. The element types might be reference types, so an array can be declared to hold nullable reference types. The following example declarations show the different syntax used to declare the nullability of the array or the elements:
```

**Pull request merged**